### ISSUE TITLE
Keep-Alive broken

### ISSUE THREAD

**Issue by kelunik** [amphp/aerys#182]
Keep alive results in hanging connections in Chrome.

 * Load one resource in a new tab.
 * Wait for `timeout` + 1, default is 6 + 1 seconds.
 * Reload page.

The result is that chrome hangs and waits for a response.

**Comment by bwoebi** [comment]
I cannot reproduce this locally.

**Comment by bwoebi** [comment]
Could you please give me more debugging detail, like what's actually happening on the TCP level (which requests go to which TCP stream - best just give me the full tcpdump sequence here) and the debug output of Aerys?

**Comment by kelunik** [comment]
```
13:17:53.167694 IP kelunik.47478 > 172.17.0.2.http-alt: Flags [P.], seq 1769:2700, ack 1047, win 252, options [nop,nop,TS val 3815678320 ecr 3396235222], length 931: HTTP: GET /get.js?... HTTP/1.1
13:17:53.211663 IP 172.17.0.2.http-alt > kelunik.47478: Flags [.], ack 2700, win 270, options [nop,nop,TS val 3396237197 ecr 3815678320], length 0
[...]
13:17:59.656198 IP 172.17.0.2.http-alt > kelunik.47504: Flags [F.], seq 1, ack 1, win 227, options [nop,nop,TS val 3396238808 ecr 3815678315], length 0
13:17:59.656538 IP 172.17.0.2.http-alt > kelunik.47506: Flags [F.], seq 3406624882, ack 3469107840, win 227, options [nop,nop,TS val 3396238808 ecr 3815678316], length 0
13:17:59.659689 IP kelunik.47506 > 172.17.0.2.http-alt: Flags [.], ack 1, win 229, options [nop,nop,TS val 3815679943 ecr 3396238808], length 0
13:17:59.659704 IP kelunik.47504 > 172.17.0.2.http-alt: Flags [.], ack 2, win 229, options [nop,nop,TS val 3815679943 ecr 3396238808], length 0
13:18:03.916184 IP kelunik.47504 > 172.17.0.2.http-alt: Flags [F.], seq 1, ack 2, win 229, options [nop,nop,TS val 3815681007 ecr 3396238808], length 0
13:18:03.916218 IP 172.17.0.2.http-alt > kelunik.47504: Flags [.], ack 2, win 227, options [nop,nop,TS val 3396239873 ecr 3815681007], length 0
13:18:03.916294 IP kelunik.47506 > 172.17.0.2.http-alt: Flags [F.], seq 1, ack 1, win 229, options [nop,nop,TS val 3815681007 ecr 3396238808], length 0
13:18:03.916307 IP 172.17.0.2.http-alt > kelunik.47506: Flags [.], ack 2, win 227, options [nop,nop,TS val 3396239873 ecr 3815681007], length 0
13:18:57.004574 IP kelunik.47478 > 172.17.0.2.http-alt: Flags [F.], seq 2700, ack 1047, win 252, options [nop,nop,TS val 3815694279 ecr 3396237197], length 0
13:18:57.047650 IP 172.17.0.2.http-alt > kelunik.47478: Flags [.], ack 2701, win 270, options [nop,nop,TS val 3396253156 ecr 3815694279], length 0
```

I've removed the communication that the request handler does with an external API.

**Comment by bwoebi** [comment]
172.17.0.2.http-alt > kelunik.47478

^ not sure why there's no FIN flag sent to that port?

Also, there's just one request in there, not two? I wanted to know how it correlates time-wise. (i.e. is chrome trying to send a request after the server already sent a FIN? Did the 6 seconds elapse since request start? request end? Or what does Chrome apply the keepalive to?)

**Comment by kelunik** [comment]
Ok, here's a dump with two requests, one is the `favicon.ico` and the other is `get.js` a few seconds later.

```
13:26:30.112622 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [S], seq 66863096, win 29200, options [mss 1460,sackOK,TS val 3824619884 ecr 0,nop,wscale 7], length 0
13:26:30.112708 IP 172.17.0.2.http-alt > kelunik.36786: Flags [S.], seq 1936990110, ack 66863097, win 28960, options [mss 1460,sackOK,TS val 3405178751 ecr 3824619884,nop,wscale 7], length 0
13:26:30.112750 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [.], ack 1, win 229, options [nop,nop,TS val 3824619885 ecr 3405178751], length 0
13:26:30.112770 IP kelunik.36790 > 172.17.0.2.http-alt: Flags [S], seq 1934885955, win 29200, options [mss 1460,sackOK,TS val 3824619885 ecr 0,nop,wscale 7], length 0
13:26:30.112787 IP 172.17.0.2.http-alt > kelunik.36790: Flags [S.], seq 2239226397, ack 1934885956, win 28960, options [mss 1460,sackOK,TS val 3405178751 ecr 3824619885,nop,wscale 7], length 0
13:26:30.112802 IP kelunik.36790 > 172.17.0.2.http-alt: Flags [.], ack 1, win 229, options [nop,nop,TS val 3824619885 ecr 3405178751], length 0
13:26:30.113074 IP kelunik.36794 > 172.17.0.2.http-alt: Flags [S], seq 3980869511, win 29200, options [mss 1460,sackOK,TS val 3824619885 ecr 0,nop,wscale 7], length 0
13:26:30.420759 IP 172.17.0.2.http-alt > kelunik.36786: Flags [P.], seq 1:880, ack 932, win 241, options [nop,nop,TS val 3405178828 ecr 3824619888], length 879: HTTP: HTTP/1.1 200 OK
13:26:30.420800 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [.], ack 880, win 242, options [nop,nop,TS val 3824619962 ecr 3405178828], length 0
13:26:30.958811 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [P.], seq 932:1769, ack 880, win 242, options [nop,nop,TS val 3824620096 ecr 3405178828], length 837: HTTP: GET /favicon.ico HTTP/1.1
13:26:30.958844 IP 172.17.0.2.http-alt > kelunik.36786: Flags [.], ack 1769, win 256, options [nop,nop,TS val 3405178962 ecr 3824620096], length 0
13:26:30.959466 IP 172.17.0.2.http-alt > kelunik.36786: Flags [P.], seq 880:1307, ack 1769, win 256, options [nop,nop,TS val 3405178962 ecr 3824620096], length 427: HTTP: HTTP/1.1 404 Not Found
13:26:30.959496 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [.], ack 1307, win 256, options [nop,nop,TS val 3824620096 ecr 3405178962], length 0
13:26:37.389366 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [P.], seq 1769:2700, ack 1307, win 256, options [nop,nop,TS val 3824621704 ecr 3405178962], length 931: HTTP: GET /get.js?... HTTP/1.1
13:26:37.432661 IP 172.17.0.2.http-alt > kelunik.36786: Flags [.], ack 2700, win 270, options [nop,nop,TS val 3405180581 ecr 3824621704], length 0
13:26:41.434478 IP kelunik.36790 > 172.17.0.2.http-alt: Flags [F.], seq 1, ack 1, win 229, options [nop,nop,TS val 3824622715 ecr 3405178751], length 0
13:26:41.434513 IP kelunik.36794 > 172.17.0.2.http-alt: Flags [F.], seq 3980869512, ack 582718826, win 229, options [nop,nop,TS val 3824622715 ecr 3405178751], length 0
13:26:41.436653 IP 172.17.0.2.http-alt > kelunik.36794: Flags [.], ack 1, win 227, options [nop,nop,TS val 3405181582 ecr 3824622715], length 0
13:26:41.436656 IP 172.17.0.2.http-alt > kelunik.36790: Flags [.], ack 2, win 227, options [nop,nop,TS val 3405181582 ecr 3824622715], length 0
13:26:53.822179 IP kelunik.36786 > 172.17.0.2.http-alt: Flags [F.], seq 2700, ack 1307, win 256, options [nop,nop,TS val 3824625812 ecr 3405180581], length 0
13:26:53.864650 IP 172.17.0.2.http-alt > kelunik.36786: Flags [.], ack 2701, win 270, options [nop,nop,TS val 3405184689 ecr 3824625812], length 0
```

**Comment by bwoebi** [comment]
I've just committed d274a1d59525ef786418c130203efb555c2ca118 - I don't think it fixes your issue, but can you please check, just to confirm that _this_ wasn't the issue?

**Comment by kelunik** [comment]
Confirming can be hard, because I can't reproduce it always, but I'll report when it happens again.

**Comment by kelunik** [comment]
Maybe this is also related to #192, but that's a big maybe.

**Comment by kelunik** [comment]
Pretty sure this is solved by 0db2343a09890c203dc829b1fe8329767f3cf75d.
